crates/sui-network/src/state_sync/server.rs

```diff
@@ -47,20 +47,38 @@ where
 
         let checkpoint = request.into_inner();
 
-        self.peer_heights
-            .write()
+        let info = match self
+            .peer_heights
+            .read()
             .unwrap()
-            .update_peer_height(peer_id, Some(checkpoint.clone()));
+            .peers
+            .get(&peer_id)
+            .copied()
+        {
+            Some(mut info) if info.on_same_chain_as_us => {
+                info.height = std::cmp::max(checkpoint.sequence_number(), info.height);
+                info
+            }
+            _ => {
+                return Ok(Response::new(()));
+            }
+        };
+
+        self.peer_heights.write().unwrap().update_peer_info(
```

Comment:
this seems likewise unsafe and in need of double-checked locking?

Reply:
same as my other comment